File: hackerrank/contests/world-codesprint-7/summing-pieces.cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long ll;
const ll mod = 1000000007;

ll pow2[1000001];

void setpow2(){
	pow2[0] = 1;
	for(ll i = 1; i < 1000001; i += 1){
		pow2[i] = (pow2[i-1] * 2)%mod;
	}
}

inline ll  getpow2(ll n){
	if (n <= 0) return 1;
	return pow2[n];
}

int main(void){
	setpow2();
	ll n;
	cin >> n;
	ll freq[n];
	
	freq[0] = getpow2(n) - 1;
	ll first = n-2, second = 0, ind = 1;
	while(first > second){
		freq[ind++] = getpow2(first--) - getpow2(second++);
	}
	for(int i = 1; i < (n+1)/2; i += 1){
		freq[i] += freq[i-1];
	}
	ll ans = 0, temp;
	for(int i = 0; i < n; i += 1){
		cout << (i < (n+1)/2 ? freq[i] : freq[n-i-1]) <<" ";
	}
	cout << endl;
	for(ll i = 0; i < n; i +=1){
		cin >> temp;
		ans = (ans + temp*(i < (n+1)/2 ? freq[i] : freq[n-i-1]))%mod;
	}
	cout << ans << endl;

}

File: hackerrank/contests/world-codesprint-7/summing-pieces.py
mod = 10**9 + 7
n = input()

a = map(int, raw_input().split())

freq = [0] * n

power = [1]*(n+1)
for i in xrange(1,n+1):
	power[i] = (power[i-1]*2)%mod

freq[0] = power[n] - 1
first = n-2
second = 0
ind = 1
while first >= second:
	freq[ind] = power[first] - power[second]
	ind += 1
	first -= 1
	second += 1

for i in xrange(1, (n+1)/2):
	freq[i] += freq[i-1]

ans = 0

for i in xrange(n):
	if i < (n+1)/2:
		ans = (ans + a[i] * freq[i])%mod
	else:
		ans = (ans + a[i]*freq[n-i-1])%mod

print ans